### PR TITLE
#72 authorization] implement user sign up

### PR DESCRIPTION
**Pull request by danylkoltak**
added api endpoint for users initial sign up

 
Screenshot 2024-12-18 171443

### CODE REVIEW

File: forum/users/views.py
Comment (mehalyna) [inline forum/users/views.py:1]: 1) Provide more detailed and user-friendly error messages, which can improve the experience for API consumers.
2) Ensure the view handles cases where the email is already taken with a clear error message.
3) Make sure that password validation errors are descriptive (this is already partially handled in your serializer).
4) Avoid returning sensitive user data (e.g., password) in the response.
5) Ensure the view is accessible only to unauthenticated users.
Like example:
```
from rest_framework.views import APIView
from rest_framework.response import Response
from rest_framework import status
from rest_framework.permissions import AllowAny
from .serializers import CustomUserSerializer

class RegisterUserView(APIView):
    """
    View for user registration.
    """
    permission_classes = [AllowAny]

    def post(self, request):
        # Pass the incoming data to the serializer
        serializer = CustomUserSerializer(data=request.data)
        
        # Validate the data
        if serializer.is_valid():
            # Save the new user
            user = serializer.save()

            # Return a success response with limited user data
            response_data = {
                "user_id": user.user_id,
                "email": user.email,
                "first_name": user.first_name,
                "last_name": user.last_name,
                "role": user.get_role_display(),
                "created_at": user.created_at,
            }
            return Response(response_data, status=status.HTTP_201_CREATED)
        
        # Return validation errors
        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```

Comment (mehalyna) [inline forum/users/views.py:1]: Write tests to cover the following cases:
- Valid user registration.
- Missing or invalid fields.
- Attempting to register with an already existing email.

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: forum/users/views.py
Comment (danylkoltak) [reply forum/users/views.py:1]: Implemented changes, provided logging for valid actions and errors, response arrives without sensitive data.

Comment (danylkoltak) [reply forum/users/views.py:1]: done in tests.py

File: forum/users/tests.py
Comment (mehalyna) [inline forum/users/tests.py:1]: The test cases do not account for edge conditions, such as:
- Very long names.
- Missing optional fields (e.g., role).
- Weak passwords.

Comment (mehalyna) [inline forum/users/tests.py:1]: The test cases do not verify whether the response data includes all expected fields (e.g., `email`, `first_name`).

Comment (mehalyna) [inline forum/users/tests.py:1]: No tests verify behavior for invalid input, such as:
- Excessive field lengths.
- Invalid data types (e.g., sending a list instead of a string).

File: forum/users/tests.py [hunk 1]
```diff
@@ -76,3 +79,49 @@ def test_empty_and_none_passwords(self):
             with self.subTest(password=password):
                 with self.assertRaises(ValidationError):
                     self.validator.validate(password)
+
+
+class UserRegistrationTests(APITestCase):
+    def setUp(self):
+        self.register_url = "/api/v1/users/user_register/"
```
Comment (mehalyna) [inline forum/users/tests.py:86]: The `register_url` is hardcoded, which could lead to maintenance issues if the API structure changes.
Use Django's `reverse` method for dynamic URL resolution:
```
from django.urls import reverse

def setUp(self):
    self.register_url = reverse("user-register")  # Assuming "user-register" is the name of the URL
    ...
```

File: forum/users/tests.py
Comment (mehalyna) [inline forum/users/tests.py:1]: There's no test to verify registration works when optional fields (e.g., `user_phone`, `title`) are not provided.

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: forum/users/views.py [hunk 2]
```diff
@@ -1,3 +1,52 @@
...
+                    "first_name": user.first_name,
+                    "last_name": user.last_name,
+                    "role": user.get_role_display(),
+                    "created_at": user.created_at,
+                }
+                return Response(response_data, status=status.HTTP_201_CREATED)
+            
+            except Exception as e:
+                logger.error(f"Error during user registration: {e}", exc_info=True)
+                return Response(
```
Comment (ne0ascorbinka) [inline forum/users/views.py:43]: You might want to consider propagating general error further as it might be useful to also let django handle it: 
```python
try:
    ...
except APIException as e:
    logger.error(f"API error during signup: {e}")
    raise
except Exception as e:
    logger.critical(f"Unexpected error during signup: {e}", exc_info=True)
    raise
```
Unless you are also handling some specific error, info about which might be informative to user:
```python
try:
    user = serializer.save()
except IntegrityError as e:
    logger.warning(f"Signup failed due to integrity error: {e}")
    return Response(
        {"detail": "A user with this email already exists."},
        status=status.HTTP_400_BAD_REQUEST,
    )
```

**Review by maksym-derkach — APPROVED** [review]

File: forum/users/tests.py
Comment (danylkoltak) [reply forum/users/tests.py:1]: done

Comment (danylkoltak) [reply forum/users/tests.py:1]: response includes first_name, last_name, email and password fields

Comment (danylkoltak) [reply forum/users/tests.py:1]: max length for first_name and last_name is 30 characters

Comment (danylkoltak) [reply forum/users/tests.py:1]: done

File: forum/users/views.py (hunk 2 above)
Comment (danylkoltak) [reply forum/users/views.py]: adjusted logger and added dynamic __name__

**Review by mehalyna — APPROVED** [review]

**Review by ne0ascorbinka — APPROVED** [review]
